### PR TITLE
Better override for perform method on partial definition

### PR DESCRIPTION
**Pull request by BrunoBernardino**
Allowing the complete override of `perform` and `authentication.test` methods when applying a partial definition, to prevent confusion of "what to execute" if there's a
`url` and a `source`, for example, or a function and a `url` in those.

Added tests to confirm the behavior.

This was noticed as part of https://github.com/zapier/zapier/pull/19935

### CODE REVIEW

File: src/tools/create-lambda-handler.js [hunk 1]
```diff
@@ -26,7 +26,48 @@ const extendAppRaw = (base, extension) => {
...
+    }
+
+    const topLevelKeys = ['triggers', 'creates', 'searches'];
+    topLevelKeys.forEach(topLevelKey => {
+      if (!_.get(_extension, [topLevelKey])) {
+        return;
+      }
+
+      Object.keys(_extension[topLevelKey]).forEach(extensionTopLevelKey => {
+        if (
```
Comment (BrunoBernardino) [inline src/tools/create-lambda-handler.js]: I think prettier actually made this more difficult to read, but 🤷‍♂️

**Review by eliangcs — COMMENTED** [review]
Pulled and tested. The change works perfectly. I have two #suggestions.

1. I wonder if we could make the code much cleaner by moving all the merge logic into the `customizer` of `_.mergeWith`. Something naive would be:

```js
const extendAppRaw = (base, extension) => {
  const customMerge = (objValue, srcValue, key) => {
    if (Array.isArray(objValue) && Array.isArray(srcValue)) {
      return objValue.concat(srcValue);
    }

    if (
      // Do full replacement when it comes to "authentication.test" and "perform"
      (key === 'perform' || key === 'test') &&
      _.isPlainObject(srcValue) &&
      _.isPlainObject(objValue)
    ) {
      return srcValue;
    }

    return undefined;
  };
  return _.mergeWith(base, extension, customMerge);
};
```

This is not very accurate, but it passes your tests. This `customMerge` tells if a value is an `authentication.test` or `(creates|searches|triggers).KEY.operation.perform` by only checking if the one last key is `test` or `perform`. Maybe we want to look closer into `srcValue` and `objValue` to see if they look like a `RequestSchema`.

2. Those diffs that remove `^`s from `package-lock.json` seem to suggest that you're using an older version of `npm`. But I'm not sure. Can you confirm? If it does come from an older version of `npm`, you'll need to upgrade `npm` and rerun `npm install` to get rid of those noises.

**Comment by BrunoBernardino** [comment]
Thank you so much @eliangcs !

1. That makes sense. The only potential problem there is if a trigger/create/search is named `perform` or `test`, where it will lose data on the final definition, and that can be very problematic. I can totally do that simpler version if we acknowledge that and are OK with it since it's probably a very edge-case.

2. I'm on `npm v5.8.0` for `node v8.11.1`. Is that outdated here?

**Comment by eliangcs** [comment]
@BrunoBernardino 

1. Yeah, so besides the one last key, we should also check if there're fields like `source`, `url`, `method`, `params`, etc to make sure we're not replacing the wrong things. We can use `zapier-platform-schema` to do the check:

```js
const RequestSchema = require('zapier-platform-schema/lib/schemas/RequestSchema');
const FunctionSchema = require('zapier-platform-schema/lib/schemas/FunctionSchema');

const isRequestOrFunction = obj => {
  return (
    RequestSchema.validate(obj).valid || FunctionSchema.validate(obj).valid
  );
};

const extendAppRaw = (base, extension) => {
  const customMerge = (objValue, srcValue, key) => {
    if (Array.isArray(objValue) && Array.isArray(srcValue)) {
      return objValue.concat(srcValue);
    }

    if (
      // Do full replacement when it comes to "authentication.test" and "perform"
      (key === 'perform' || key === 'test') &&
      _.isPlainObject(srcValue) &&
      _.isPlainObject(objValue) &&
      isRequestOrFunction(srcValue) &&
      isRequestOrFunction(objValue)
    ) {
      return srcValue;
    }

    return undefined;
  };
  return _.mergeWith(base, extension, customMerge);
};
```

2. I'm using npm 6.1.0. I believe npm 6 started using loose versions (vs exact versions in npm 5) in package-lock.json. See https://github.com/npm/npm/issues/20434.

**Comment by BrunoBernardino** [comment]
Alright @eliangcs please check now!

File: src/tools/create-lambda-handler.js [hunk 2]
```diff
@@ -19,11 +19,39 @@ const environmentTools = require('./environment');
...
 const extendAppRaw = (base, extension) => {
-  const concatArray = (objValue, srcValue) => {
+  const keysToOverride = [
+    'test',
+    'perform',
+    'performList',
+    'performSubscribe',
+    'performUnsubscribe'
+  ];
+  const concatArray = (objValue, srcValue, key) => {
```
Comment (eliangcs) [inline src/tools/create-lambda-handler.js]: We should probably rename `concatArray` to something else as the function doesn't only do array concatenation anymore.

**Review by eliangcs — APPROVED** [review]
👍Looks good! Once you rename that `concatArray` function, I think you're good to go.
